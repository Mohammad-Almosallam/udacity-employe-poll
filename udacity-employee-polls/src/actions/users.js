export const RECEIVE_USERS = "RECEIVE_USERS";
export const SUBMIT_USER_ANSWER = "SUBMIT_USER_ANSWER";
export const SUBMIT_USER_NEW_QUESTION = "SUBMIT_USER_NEW_QUESTION";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}
export function submitUserAnswer({ qid, authedUser, answer }) {
  return {
    type: SUBMIT_USER_ANSWER,
    qid,
    authedUser,
    answer,
  };
}

export function submitUserNewQuestion(info) {
  const question = info.question;
  return {
    type: SUBMIT_USER_NEW_QUESTION,
    author: question.author,
    qId: question.id,
  };
}
