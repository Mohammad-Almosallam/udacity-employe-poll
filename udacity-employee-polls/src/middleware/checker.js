const checker = (store) => (next) => (action) => {};

export default checker;
