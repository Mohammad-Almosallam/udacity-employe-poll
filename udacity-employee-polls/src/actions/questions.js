import { saveQuestionAnswer, saveQuestion } from "../utils/api";
import { submitUserAnswer, submitUserNewQuestion } from "./users";
import { showLoading, hideLoading } from "react-redux-loading-bar";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const SUBMIT_QUESTION_ANSWER = "SUBMIT_QUESTION_ANSWER";
export const SUBMIT_NEW_QUESTION = "SUBMIT_NEW_QUESTION";

// -------------------- RECEIVING ALL QUESTIONS --------------------

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}
// -----------------------------------------------------------------

// -------------------- SUBMITTING NEW QUESTIONS --------------------
export function submitNewQuestion(question) {
  return {
    type: SUBMIT_NEW_QUESTION,
    question,
  };
}

export function handleSubmitNewQuestion(info) {
  return (dispatch) => {
    dispatch(showLoading());
    return saveQuestion(info)
      .then((question) => dispatch(submitNewQuestion(question)))
      .then((question) => dispatch(submitUserNewQuestion(question)))
      .then(() => dispatch(hideLoading()));
  };
}
// -------------------------------------------------------------------

// -------------------- SUBMITTING ANSWER FOR A QUESTION --------------------
export function submitQuestionAnswer({ qid, authedUser, answer }) {
  return {
    type: SUBMIT_QUESTION_ANSWER,
    qid,
    authedUser,
    answer,
  };
}

export function handleSubmitQuestionAnswer(info) {
  return (dispatch) => {
    dispatch(showLoading());

    return saveQuestionAnswer(info)
      .then(() => dispatch(submitQuestionAnswer(info)))
      .then(() => dispatch(submitUserAnswer(info)))
      .then(() => dispatch(hideLoading()));
  };
}
// --------------------------------------------------------------------------
