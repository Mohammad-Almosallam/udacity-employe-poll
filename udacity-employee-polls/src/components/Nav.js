import React from "react";
import { Flex, Box, Text, Button, Image, useColorMode } from "@chakra-ui/react";
import { Link } from "react-router-dom";
import {
  IoLogOutOutline,
  IoSunnyOutline,
  IoMoonOutline,
} from "react-icons/io5";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";

function Nav(props) {
  const { colorMode, toggleColorMode } = useColorMode();

  return (
    <Box>
      <Flex alignItems={"center"} p={"2rem"} justifyContent={"space-between"}>
        <Flex gap={"2rem"}>
          <Link to="/">Home</Link>
          <Link to="/leaderboard">Leaderboard</Link>
          <Link to="/add">New</Link>
        </Flex>
        <Flex alignItems={"center"} gap={"2rem"}>
          <Flex alignItems={"center"} gap={"1rem"}>
            <Image src={props.avatarURL} w={"50px"} h={"50px"} />
            <Text>{props.loggedInUser}</Text>
          </Flex>
          <Button
            bg={"transparent"}
            borderRadius={"50%"}
            width="40px"
            p={0}
            _hover={{ backgroud: "transparent" }}
            onClick={toggleColorMode}
          >
            {colorMode === "light" ? <IoMoonOutline /> : <IoSunnyOutline />}
          </Button>
          <Button
            colorScheme={""}
            gap={"0.4rem"}
            onClick={() => {
              props.dispatch(setAuthedUser(""));
            }}
            backgroundColor={"black"}
            color={"white"}
          >
            Logout
            <IoLogOutOutline fontSize={"1.3rem"} />
          </Button>
        </Flex>
      </Flex>
    </Box>
  );
}

const mapStateToProps = ({ authedUser, users }) => ({
  loggedInUser: authedUser,
  avatarURL:
    authedUser !== null
      ? Object.values(users).find((users) => users.id === authedUser).avatarURL
      : null,
});

export default connect(mapStateToProps)(Nav);
