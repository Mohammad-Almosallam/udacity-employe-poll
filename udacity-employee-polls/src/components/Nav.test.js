import React from "react";
import Nav from "./Nav";
import reducer from "../reducers";
import middleware from "../middleware";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import { setAuthedUser } from "../actions/authedUser";
import { handleInitialData } from "../actions/shared";

const store = createStore(reducer, middleware);

describe("Nav", () => {
  it("renders the Nav component completely and correctly", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
  it("should display the DarkMode button if logged-in", async () => {
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("zoshikanlu"));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(component.queryByTestId("dark-btn")).toBeInTheDocument();
  });
  it("should display the DarkMode button if logged-in", async () => {
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("zoshikanlu"));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.click(component.getByTestId("dark-btn"));
    expect(component.getByTestId("dark-btn").getAttribute("value")).toEqual(
      "dark"
    );
  });
});
