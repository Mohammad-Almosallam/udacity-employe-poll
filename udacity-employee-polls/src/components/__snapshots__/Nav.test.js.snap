// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders the Nav component completely and correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
      >
        <div
          class="css-1rgb79v"
        >
          <div
            class="css-1gippar"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              href="/add"
            >
              New
            </a>
          </div>
          <div
            class="css-1d6oiat"
          >
            <div
              class="css-rvp98e"
            >
              <img
                class="chakra-image css-1u7uhqd"
              />
              <p
                class="chakra-text css-0"
              />
            </div>
            <button
              class="chakra-button css-4wn91n"
              data-testid="dark-btn"
              type="button"
              value="dark"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 48v48m0 320v48m147.08-355.08l-33.94 33.94M142.86 369.14l-33.94 33.94M464 256h-48m-320 0H48m355.08 147.08l-33.94-33.94M142.86 142.86l-33.94-33.94"
                  fill="none"
                  stroke-linecap="round"
                  stroke-miterlimit="10"
                  stroke-width="32"
                />
                <circle
                  cx="256"
                  cy="256"
                  fill="none"
                  r="80"
                  stroke-linecap="round"
                  stroke-miterlimit="10"
                  stroke-width="32"
                />
              </svg>
            </button>
            <button
              class="chakra-button css-yo5twi"
              type="button"
            >
              Logout
              <svg
                fill="currentColor"
                font-size="1.3rem"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40m64 160l80-80-80-80m-192 80h256"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="32"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
    >
      <div
        class="css-1rgb79v"
      >
        <div
          class="css-1gippar"
        >
          <a
            href="/"
          >
            Home
          </a>
          <a
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            href="/add"
          >
            New
          </a>
        </div>
        <div
          class="css-1d6oiat"
        >
          <div
            class="css-rvp98e"
          >
            <img
              class="chakra-image css-1u7uhqd"
            />
            <p
              class="chakra-text css-0"
            />
          </div>
          <button
            class="chakra-button css-4wn91n"
            data-testid="dark-btn"
            type="button"
            value="dark"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 48v48m0 320v48m147.08-355.08l-33.94 33.94M142.86 369.14l-33.94 33.94M464 256h-48m-320 0H48m355.08 147.08l-33.94-33.94M142.86 142.86l-33.94-33.94"
                fill="none"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="32"
              />
              <circle
                cx="256"
                cy="256"
                fill="none"
                r="80"
                stroke-linecap="round"
                stroke-miterlimit="10"
                stroke-width="32"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-yo5twi"
            type="button"
          >
            Logout
            <svg
              fill="currentColor"
              font-size="1.3rem"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40m64 160l80-80-80-80m-192 80h256"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="32"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
